File: backend/src/main/java/com/codeloom/backend/converter/AbstractJsonReadingConverter.java
package com.codeloom.backend.converter;

import org.postgresql.util.PGobject;
import org.springframework.core.convert.converter.Converter;
import tools.jackson.databind.ObjectMapper;

public abstract class AbstractJsonReadingConverter<T> implements Converter<PGobject, T> {
  private final ObjectMapper mapper;
  private final Class<T> type;

  protected AbstractJsonReadingConverter(ObjectMapper m, Class<T> t) {
    mapper = m;
    type = t;
  }

  public T convert(PGobject source) {
    return mapper.readValue(source.getValue(), type);
  }
}

File: backend/src/main/kotlin/com/codeloom/backend/converter/AbstractJsonReadingConverter.kt
package com.codeloom.backend.converter

import org.postgresql.util.PGobject
import org.springframework.core.convert.converter.Converter
import tools.jackson.databind.ObjectMapper

abstract class AbstractJsonReadingConverter<T>(
    private val objectMapper: ObjectMapper,
    private val clazz: Class<T>,
) : Converter<PGobject, T> {
    override fun convert(source: PGobject): T {
        return objectMapper.readValue(source.value, clazz)
    }
}
